src/libical/icalcomponent.c

```diff
@@ -1488,12 +1504,32 @@ void icalcomponent_set_duration(icalcomponent *comp, struct icaldurationtype v)
 struct icaldurationtype icalcomponent_get_duration(icalcomponent *comp)
 {
     icalcomponent *inner = icalcomponent_get_inner(comp);
-
-    icalproperty *end_prop = icalcomponent_get_first_property(inner, ICAL_DTEND_PROPERTY);
+    const icalcomponent_kind kind = icalcomponent_isa(inner);
+    icalproperty *end_prop;
+    struct icaltimetype end;
+
+    switch(kind) {
+    case ICAL_VAVAILABILITY_COMPONENT:
+    case ICAL_VEVENT_COMPONENT:
+    case ICAL_XAVAILABLE_COMPONENT:
+        end_prop = icalcomponent_get_first_property(inner, ICAL_DTEND_PROPERTY);
+        if (end_prop) {
+            end = icalcomponent_get_dtend(inner);
+        }
+        break;
+    case ICAL_VTODO_COMPONENT:
+        end_prop = icalcomponent_get_first_property(inner, ICAL_DUE_PROPERTY);
+        if (end_prop) {
+            end = icalcomponent_get_due(inner);
+        }
+        break;
+    default:
+        /* The libical API is used incorrectly */
+        return icaldurationtype_null_duration();
+    }
 
     icalproperty *dur_prop = icalcomponent_get_first_property(inner, ICAL_DURATION_PROPERTY);
```

Comment:
same here. `icalproperty *dur_prop` and `struct icaldurationtype ret` must be declared above the switch